src/h5cpp/attribute/attribute.hpp

```diff
@@ -355,7 +355,7 @@ void Attribute::write(const T &data) const
 template<typename T>
 void Attribute::read(T &data) const
 {
-  auto mem_type = datatype::create<T>(data);
+  auto mem_type = datatype();
   read(data,mem_type);
```

Comment:
I am not sure this is the desired behavior... why do we want to attempt a read with an abstract, undefined datatype rather than the one inferred from the type of T?

Reply:
We have to call `H5Aread()` with the proper`file_type`  (not as it is now in `master` with `mem_type`) otherwise we get errors from `H5Aread()` we cannot control. After the `H5Aread()` call we can cast the read value to the proper T type (which in the fix is done automatically by c++ from the file Trait to the T data output buffer). If it is not possible we get an exception from casting. But for string we have relation many-to-one (file-type to mem-type) types so using file_type is more logical. 

For non string types it the current `master` implementation can causes problems in reading e.g. unsigned/signed  integers.

Reply:
I get this many-to-one relation, and I totally agree that for strings we should not care what they are in the file -- they end up in our std::string regardless. So, for strings you are probably right that we should use the file_type. But it seems that with this fix it will be come the file_type for all types, not just strings? Which, I think, would actually be wrong. 

We do not really have or test any cases where we go in with our type and have HDF5 do type conversion, but it might be the case that this already functions for converting numeric types between file type and native type.

But maybe I am wrong with this reasoning?

Another place for this fix would be the read functions that decide what in particular to do with strings, where they can duplicate the file_type explicitly.